Replay of historic production Tableland events

jsign opened the pull request:
# Summary

This PR includes a new test that replays all Tableland history and checks that each chain's last Tableland state hash matches a defined expected hash.

---

This test aims to detect two main problematic situations that are hard to see via PRs:

**Detect non-deterministic execution**

If any event execution isn’t deterministic, it will create a snowball effect leading to a different state hash that won’t match the expected one.

Any validator that syncs from scratch should always arrive at the same state hash in each chain after it syncs to the tip. This test enforces this invariant.

**Detect if fixes or new features unintentionally change history**

If we make a bug fix or implement a new feature, there’s the chance that, under these new rules, the execution of historical events can lead to a different state. This is problematic since a new validator doing a cold-sync would have a different state hash than validators that simply update the version.

Today, we should expect that all fixes and features don’t change the end result of executing historical events. Of course, this will be relaxed when we implement fixes or features that activate at some particular height of chains (and forward). As in, having a multi-versioned execution pipeline that we should do eventually.

Even in the future, this test is already covering what holds true. With a multi-versioned execution pipeline, this test would properly detect that previous history is still backward compatible with new features that should activate further in time.

This topic is a bit of a rabbit hole for this PR, so I’ll stop here. We can return to this topic when we work in this multi-versioned execution pipeline; I just wanted to surface that this PR is already adding test coverage around it.

---

Before starting this PR, it felt pretty obvious that we’d already detected some problems, and that was the case:

- It detected a bug that caused a panic; this was fixed (explained in PR comments)
- It surfaced an obvious optimization that allowed us to replay the history 2x faster (explained in PR comments)

Some days ago, we had a situation where a previous version of the parser allowed some non-deterministic function calls. To double-check this test would have detected that situation, I downgraded the `sqlparser` dependency to the buggy version. I verified that this test detects a mismatch between the expected and calculated hash. More concretely, each test run produced different hashes, a signal of non-deterministic execution.

---

This test also serves as a building block for further work regarding benchmarking the validator since replaying production history might be the best way to do it. I’ll soon create a GH issue to track that work and drop some ideas on how standing on the shoulders of this test, we can discover other interesting information.

As quick information, here’re some facts:

- The test runs using an in-memory SQLite database.
- Hashing every X block is disabled; we only hash at the end since the last hash would have the snowball effect of any previous divergence. [Note about this later]
- The historical data used in the test contains an ~up-to-date version of the current history, totaling 192,770 events.
- The test without the `-race` flag (thus no runtime overhead) takes ~60s to run in my machine.

This means we could process ~3200 events per second (with an in-memory SQLite).

This is a good number to know since measuring the max throughput of our Tableland event execution logic. Of course, running with a file-based SQLite will have less throughput, but that is related to SQLite and *probably* not code in `go-tableland`. 

As part of future benchmarking work, we can run this test both in-memory and with a real SQLite database to continue understanding how adding more layers affects performance.

The quick conclusion is that probably any performance improvement won’t come from any business logic in Go since 3k events/second is more than enough for a real situation today. The bottleneck will probably live in the interaction with SQLite regarding its single-writer constraint. (Again, rabbit hole here).

# Implementation overview

The implementation is a new Go test that replays the history of all the known events, so it’s probably the complete e2e execution since it uses real data.

The following is a high-level description of how this works:

- The test relies on a `testdata/evm_history.db.zst` file, a zstandard compressed SQLite database containing a `system_evm_events` extracted from a real production database.
- It spins up the complete execution stack for all chains and executes all the events until it reaches the tip.
- For the above to work, the `EventFeed` now also accepts a new `sqlitechainclient`, another implementation of an Ethereum API. Apart from the Alchemy/Infura backend, you can “plug” this new component with the same API but provides block headers and logs backed by SQLite. The `EventFeed` is agnostic on how it receives events. (More in PR comments)
- After we detect that all chains are synced, the state hash is calculated for each chain and compared to fixed values which are the source of truth.

Notes:

- The test automatically detects all the existing chains in history and creates the appropriate execution stack. If we support new chains in the future, it’d be automatically detected.
- There’s a new `make generate-history-db` that helps to update the `testdata/evm_history.db.zst` based on a new `database.db`. In the future, we should keep updating the historical events to keep “up to date” throughout the continuously built history.
- The uncompressed SQLite database is ~200MiB, which is too big to commit to Git. The compressed size is ~23MiB, which is much better. The test automatically decompresses the database, uses it to run it, and is later discarded.

As mentioned before, running this test in my machine without `-race` takes ~60s. With `-race` takes ~140s. This was relevant to plan on what to do for CI, which I decided to work this way:

- This new “run history” test only runs if you don’t provide the `-short` flag, like `go test -short`.
- In CI, I added an extra dimension to the test matrix, so at the same time, we’re running all the tests but this one in a job and the “replay history” test in a separate job that runs in parallel. (See the new "Test / test (1.19.x, ubuntu-latest, test-replayhistory)" item in the actions summary)
- This allows for overlapping both runs, so we minimize adding minutes to the complete execution of CI (see PR comments for more details).
- In CI, we run this new long test with `-race`. It might take longer, but since it's syncing all the events in all the chains as fast as possible, it's indirectly a great test case to detect potential dataraces.

# Implementation details and review orientation

[WIP - After creating PR comments, I’ll add here some suggestions about the review order of files]

sanderpick commented:
Great PR description as always. Nice idea with `sqlitechainclient` impl. Would hashing every X block help pin point where some indeterminism arrises? Or is that tradeoff not worth the extra overhead?

jsign commented:
> Great PR description as always. Nice idea with `sqlitechainclient` impl. Would hashing every X block help pin point where some indeterminism arrises? Or is that tradeoff not worth the extra overhead?

To pin point a specific block we'd have to run the hashes for every block, and that's quite heavy. For this test, I want to detect that there's a problem _as fast as possible. So if the final hash differs, we know there's a problem somewhere. Since that situation should only happen rarely, I prefer the test to run as fast as possible.

But if we find a case where we detected a problem, we can do further work based on this test to, now that we know there's some hairy thing going on, run the hashes ever X blocks and start narrowing down _where_ is the problem.

jsign commented:
As an extra comment, if we'd the feature of only calculating hashes of tables that changes since the previous calculation maybe that would make this less heavy. So we'll see if at some point it can be "cheap enough" to make it the default to find where it is.

sanderpick commented:
Makes sense 💯

brunocalza reviewed (approved):
Outstanding! So good to see `EventFeed` untouched. Really nice implementation :)

jsign commented on Makefile:
```diff
@@ -42,6 +42,13 @@ clean-mocks:
 	rm -rf mocks
 .PHONY: clean-mocks
 
+EVM_EVENTS_ORIGIN:="docker/deployed/testnet/api/backup_database.db"
+EVM_EVENTS_TARGET:="pkg/eventprocessor/impl/testdata/evm_history.db"
+generate-history-db:
+	rm -f ${EVM_EVENTS_TARGET}
+	sqlite3 ${EVM_EVENTS_ORIGIN} 'ATTACH DATABASE ${EVM_EVENTS_TARGET} as target' 'CREATE TABLE target.system_evm_events as select * from system_evm_events'
+	zstd -f ${EVM_EVENTS_TARGET}
+	rm ${EVM_EVENTS_TARGET}
```
Nope, it isn't retrieved by `scp`. In the cloud validator we don't have a full history of events, since that validator started persisted events when the feature came into existence and forward, so it'd have events missing.

The database I used was from my local validator, which I synced from scratch after the PR saving events was created, so AFAIK is the only one that has all the events.

Whenever we do a cold sync of the cloud validator, yes, we should retrieve it with `scp` and use that.

Reg the datetime, we could but technically speaking the `block_number` of the events already give a timestamp of the events in the chains. Or maybe you were thinking about date-times for other reason?

brunocalza replied:
The datetime was mostly for us to visually know the date of the backup. But I was assuming it came from a remote backup. But makes sense it was from your local machine.

jsign commented on cmd/api/main.go:
```diff
@@ -76,6 +80,18 @@ func main() {
 		log.Fatal().Err(err).Msg("instrumenting sql validator")
 	}
 
+	executorsDB, err := otelsql.Open("sqlite3", databaseURL)
+	if err != nil {
+		log.Fatal().Err(err).Msg("creating executors db")
+	}
+	executorsDB.SetMaxIdleConns(0)
+	executorsDB.SetMaxOpenConns(1)
+	if err := otelsql.RegisterDBStatsMetrics(
+		executorsDB,
+		otelsql.WithAttributes(attribute.String("name", "executors"))); err != nil {
+		log.Fatal().Err(err).Msg("registering executors db stats")
+	}
```
Oh yep, typo in the comment. The code is correct.

jsign commented on pkg/eventprocessor/eventfeed/impl/sqlitechainclient/sqlitechainclient.go:
```diff
@@ -0,0 +1,161 @@
+			  where chain_id=?1 and 
+			        block_number between ?2 and ?3 and
+					address=?4
+			  order by block_number asc`
```
Oh, good point, it should be added. I'll push a commit with that.

jsign commented on pkg/eventprocessor/eventfeed/impl/sqlitechainclient/sqlitechainclient.go:
```diff
@@ -0,0 +1,161 @@
+	return logs, nil
+}
+
+// HeaderByNumber returns the block header of the chain.
+// Note that it can only be called with block==nil (i.e: last known block) since the underlying SQLite database isn't
+// a full replication of the underlying chain.
+func (scc *SQLiteChainClient) HeaderByNumber(ctx context.Context, block *big.Int) (*types.Header, error) {
+	if block != nil {
+		return nil, errors.New("the current implementation only allows returning the latest block number")
+	}
+
+	scc.onceChainTip.Do(func() {
+		blockNumber, err := scc.getChainTipBlockNumber(ctx)
+		if err != nil {
+			scc.log.Error().Err(err).Msg("loading chain tip block number")
+			scc.chainTipBlockNumber = -1
+			scc.onceChainTip = sync.Once{} // Reset to retry in the next `HeaderByNumber(...)` call
+			return
+		}
+		scc.chainTipBlockNumber = blockNumber
+	})
+	if scc.chainTipBlockNumber == -1 {
+		return nil, fmt.Errorf("chain tip block number couldn't be loaded")
+	}
+
+	return &types.Header{
+		Number: big.NewInt(scc.chainTipBlockNumber),
+	}, nil
+}
```
Exactly.

jsign commented on pkg/eventprocessor/impl/eventprocessor.go:
```diff
@@ -278,19 +278,22 @@ func (ep *EventProcessor) executeBlock(ctx context.Context, block eventfeed.Bloc
 }
 
-func (ep *EventProcessor) calculateHash(ctx context.Context, bs executor.BlockScope) error {
+func (ep *EventProcessor) calculateHash(ctx context.Context, bs executor.BlockScope) (string, error) {
```
Oh, yeah, I'll simplify it. I think I did this API change before realizing that (unfortunately) I needed to create the block scope manually, so I could call that directly.
